### DISCUSSION TITLE
ironwail on macOS / MGL suggestion (OpenGL 4.6 to Metal mapping)

### DISCUSSION THREAD

**Discussion by bazilio-ua** [andrei-drexler/ironwail#114] [Ideas]
Hi @andrei-drexler, 
It would be good to run ironwail on macOS if this is possible using these tool: 
https://github.com/openglonmetal/MGL
https://www.reddit.com/r/opengl/comments/qtxgph/its_alive_opengl_45_on_mac_os_metal/

How do you think this is possible and will it make sense?
Thanks!

**Comment by andrei-drexler** [comment]
That does look like a very promising project, but I'm not sure if it's ready yet. It might already have all the necessary functionality, but I don't own/have access to a Mac, so I couldn't test/debug this myself anyway. Longer term, though, I'd actually like to add native support for multiple backends, like Vulkan, and maybe Metal if/when I do get my hands on a Mac.

**Reply by eierfrucht** [reply]
Check out how GZDoom or VkQuake or even Raze for that matter all handle Metal. They have a Vulkan backend that is further handled by a third party lib called MoltenVK down the pipeline. Works beautifully on macos. All you need for Mac is vulkan support, the next step of integrating MoltenVK ain't no rocket surgery. Supporting metal natively or sticking in a gl—>metal translation layer is just silly. With native Vulkan support you get the thing running both on modern Linux and macos. GZDoom devs barely have any Mac computers to test with, yet their Vulkan—>MoltenVK solution has worked beautifully for Mac users since some years... For Jesus’ sake add Vulkan support then suddenly Ironwail works on Mac and cutting edge Linux.

OGL is largely deprecated and only makes sense for running on derelict hardware. Vulkan is natively supported even on old OSes like Windows 7, it’s in the drivers for decade+ old GPUs, it has become the de facto modern standard for Linux, and is very easily translated into Metal with the help of MoltenVK.

Rebasing ironwail around vulkan would make it truly cross-platform with zero need for catering to more than one rendering API. A modern Quake source port without Vulkan support sounds a bit... baffling.

**Reply by eierfrucht** [reply]
P.S. VkQuake development has been officially abandoned in no small way due to people mass migrating over to Ironwail and other ports, chances are the former VkQuake dev would be happy to know that his Vulkan-related body of work continued to live on as part of Ironwail. Both engines descend from quakespasm so copy pasting the vulkan stuff from vkquake could work? And VkQuake runs on Mac via MoltenVK without issues.
